shared-bindings/gnss/GNSS.c

```diff
@@ -0,0 +1,245 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/gnss/GNSS.h"
+#include "shared-bindings/util.h"
+
+#include "py/objproperty.h"
+#include "py/runtime.h"
+
+//| class GNSS:
+//|     """Get updated positioning information from GNSS
+//|
+//|     Usage::
+//|
+//|         import gnss
+//|         import time
+//|
+//|         gps = gnss.GNSS()
+//|         gps.select(gnss.SatelliteSystem.GPS)
+//|         gps.start()
+//|         last_print = time.monotonic()
+//|         while True:
+//|             gps.update()
+//|             current = time.monotonic()
+//|             if current - last_print >= 1.0:
+//|                 last_print = current
+//|                 if gps.fix is gnss.PositionFix.INVALID:
+//|                     print("Waiting for fix...")
+//|                     continue
+//|                 print("Latitude: {0:.6f} degrees".format(gps.latitude))
+//|                 print("Longitude: {0:.6f} degrees".format(gps.longitude))"""
+//|
+
+//|     def __init__(self, ):
+//|         """Turn on the GNSS."""
+//|         ...
+//|
+STATIC mp_obj_t gnss_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    gnss_obj_t *self = m_new_obj(gnss_obj_t);
+    self->base.type = &gnss_type;
+
+    common_hal_gnss_construct(self);
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|     def deinit(self, ) -> Any:
+//|         """Turn off the GNSS."""
+//|         ...
+//|
+STATIC mp_obj_t gnss_obj_deinit(mp_obj_t self_in) {
+    gnss_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_gnss_deinit(self);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(gnss_deinit_obj, gnss_obj_deinit);
+
+STATIC void check_for_deinit(gnss_obj_t *self) {
+    if (common_hal_gnss_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+//|     def select(self, system: gnss.SatelliteSystem) -> Any:
+//|         """Add specified satellite system to selection for positioning.
+//|
+//|         :param gnss.SatelliteSystem system: satellite system to use"""
+//|         ...
+//|
+STATIC mp_obj_t gnss_obj_select(mp_obj_t self_in, mp_obj_t system_obj) {
+    gnss_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    check_for_deinit(self);
+
+    gnss_satellitesystem_t system = gnss_satellitesystem_obj_to_type(system_obj);
+    common_hal_gnss_select(self, system);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(gnss_select_obj, gnss_obj_select);
+
+//|     def deselect(self, system: gnss.SatelliteSystem) -> Any:
```

Comment:
Why would someone select or deselect a system after initialization?

Reply:
In most cases, the user will use the `select` function after creating the GNSS object to select satellite systems. The selection is not in the constructor because we do not know how many satellite systems the user will choose.

The `deselect` function is added because since we can select something, we should also be able to deselect it. I can remove this function if you think it is unnecessary.